csrc/dynamic_transform.cpp

```diff
@@ -154,6 +181,97 @@ class DynamicTransformInitialInfoBuilder : public IterVisitor {
   std::vector<Val*> leaf_dynamic_vals_;
 };
 
+//! This performs a depth-first search from outputs toward inputs for empty
+//! tensors. It does not traverse past any zero tensors it finds; this is why
+//! this is implemented as a single function instead of with BackwardVisitor.
+//! Additionally, we check inputs since they might actually be disconnected from
+//! outputs.
+std::vector<EmptyTensorDescriptor> findEmptyTensors(
+    ExpressionEvaluator* expr_eval) {
+  auto fusion = FusionGuard::getCurFusion();
+  std::vector<EmptyTensorDescriptor> empty_tensors;
+  std::vector<Val*> vals(fusion->inputs());
+  vals.insert(vals.end(), fusion->outputs().begin(), fusion->outputs().end());
+  std::unordered_set<TensorView*> visited;
+
+  while (!vals.empty()) {
+    auto val = vals.back();
+    vals.pop_back();
+    if (!val->isA<TensorView>()) {
+      continue;
+    }
+    auto tv = val->as<TensorView>();
+    if (visited.find(tv) != visited.end()) {
+      continue;
+    }
+    visited.insert(tv);
+
+    std::vector<size_t> empty_axes;
+    auto rfactor = TensorDomain::noReductions(tv->getMaybeRFactorDomain());
```

Comment:
Suppose a tensor has some size-zero dimensions. If any of those are not reduced, then the output will have some zero dimensions. In that case, the backward pass to find empty tensors will identify them, or their uses and the reduction will become dead code.

If all empty dimensions are reduced, then the output will not be empty, so this will be the furthest downstream empty tensor, meaning we're handling it directly. In `DynamicTransformConcretizer::removeEmptyBranches()`, we replace the tensor with a `full()`, then we look at its `uses()`. There we handle reductions, Welfords, pads, and cats. In the case of a reduction, we replace the reduction with a `full(init_value)`.

Reply:
I was a little confused here because I considered `tv` as a reduction output, so ignoring reduction domains didn't make sense. Looks like we are using the axis information collected from reduction inputs, so doing `noReductions` makes sense.